### PR TITLE
Only store wanted data in offline mode

### PR DESCRIPTION
**Pull request by johnbotris**
we still write everything to the offline storage, and just delete it on login if it is outside the range, or not matching criteria. The database is also compacted at this time.

Offline Storage is no longer generic, at least in the cleanup phase, it now knows about mail and mailfolders and a couple other types. This might have to do for now, unless we just want to pull out the cleanup logic into some other wrapper, but that seems like useless indirection.

I opted to store the time range in the offline db itself. at first i kept it in the device config, but it's needed in both 
main and worker, and it's needed sooner in the worker so it is simpler to do this way.

time range exclusion is also opt-in for types, since most entities are likely to be rather old (e.g. MailFolder, Contact) but we still would want to always keep them cached. It's really only mail related types that we want to be clearing out at all as far as I can tell.

#3992

### CODE REVIEW

File: src/desktop/db/OfflineDb.ts [hunk 1]
```diff
@@ -137,7 +137,7 @@ export class OfflineDb {
 	getIdsInRange(type: string, listId: Id): Array<Id> {
 		const range = this.getRange(type, listId)
 		if (range == null) {
-			throw new Error(`no range exists for ${type} and list ${listId}`)
+			return []
```
Comment (charlag) [inline src/desktop/db/OfflineDb.ts:140]: Is there a reason to change this?

File: src/desktop/db/OfflineDb.ts [hunk 2]
```diff
@@ -171,6 +171,24 @@ export class OfflineDb {
 		return result?.entity ?? null
 	}
 
+	getAllOfType(type: string): Array<Buffer> {
+
+		// We query both entity tables, but at most one of the queries should ever actually return data
```
Comment (charlag) [inline src/desktop/db/OfflineDb.ts:176]: I would rather do one of:
 - make two methods
 - add a parameter

File: src/desktop/db/OfflineDb.ts [hunk 3]
```diff
@@ -207,6 +225,30 @@ export class OfflineDb {
 			.run({key, value})
 	}
 
+	deleteEntitiesBeforeId(type: string, cutoffId: Id) {
```
Comment (charlag) [inline src/desktop/db/OfflineDb.ts:228]: When we call it we also know whether it's list or not

File: src/settings/MailSettingsViewer.ts [hunk 4]
```diff
@@ -324,6 +331,27 @@ export class MailSettingsViewer implements UpdatableSettingsViewer {
...
+		if (!isOfflineStorageAvailable()) {
+			return null
+		}
+
+		return m(TextFieldN, {
+			label: "storedDataTimeRange_label",
+			value: stream(lang.get("numDays_label", { "{numDays}": this.offlineStorageWrapper.getTimeRange()})),
+			disabled: true,
+			injectionsRight: () => [m(ButtonN, {
+				// TODO Translate
```
Comment (charlag) [inline src/settings/MailSettingsViewer.ts:344]: remove

File: src/settings/OfflineStorageSettings.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import {TextFieldN, TextFieldType} from "../gui/base/TextFieldN.js";
+import stream from "mithril/stream";
+import * as cborg from "cborg"
+
+export class OfflineStorageSettingsModel {
+
+    private _timeRange = OFFLINE_STORAGE_DEFAULT_TIME_RANGE_DAYS
+
+    async isOfflineStorageEnabled(): Promise<boolean> {
+        return isOfflineStorageAvailable() && locator.loginFacade.isPersistentSession()
```
Comment (charlag) [inline src/settings/OfflineStorageSettings.ts:18]: inject `loginFacade`

File: src/settings/OfflineStorageSettings.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            }
+        }
+    }
+
+    private userId(): Id {
+        return logins.getUserController().userId
+    }
+
+    private db(): OfflineDbFacade {
+        return locator.offlineDbFacade
```
Comment (charlag) [inline src/settings/OfflineStorageSettings.ts:51]: and here

File: src/settings/OfflineStorageSettings.ts (hunk 5 above)
Comment (charlag) [reply src/settings/OfflineStorageSettings.ts:18]: Actually just use `UserController` for that

File: src/settings/OfflineStorageSettings.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+import {logins} from "../api/main/LoginController.js";
+import {OfflineDbFacade} from "../desktop/db/OfflineDbFacade.js";
+import {defer} from "@tutao/tutanota-utils";
+import {Dialog} from "../gui/base/Dialog.js";
+import m from "mithril";
+import {TextFieldN, TextFieldType} from "../gui/base/TextFieldN.js";
+import stream from "mithril/stream";
+import * as cborg from "cborg"
+
+export class OfflineStorageSettingsModel {
```
Comment (charlag) [inline src/settings/OfflineStorageSettings.ts:13]: TODO free/premium range

Comment (charlag) [reply src/settings/OfflineStorageSettings.ts:13]: TODO: plural translation
TODO: negative

**Review by charlag — COMMENTED** [review]
Two issues with mail lists:
 * We forget whether we loaded the full list because we unconditionally shrink ranges. We do not need to do that if there's nothing to delete.
 * We can't load the cached part of the list if it's less than what UI is requesting because cache will try to get as many elements as was requested from network and it will just fail. We need to somehow both return mails and communicate that we couldn't load a part of it (or that there's more items)

File: src/desktop/db/OfflineDb.ts (hunk 1 above)
Comment (johnbotris) [reply src/desktop/db/OfflineDb.ts:140]: it was just for some tests but it's not necessary

**Comment by johnbotris** [comment]
Loaded data will always be stored in the offline cache, and is cleared upon login
### Test Notes
- [ ] When offline mode is not enabled, no "Local data" section is shown in "Email Settings"

Test these with offline mode enabled (on desktop, login for the first time, go to Desktop settings, enable offline mode, log out and back in):
- [ ] A free user is not allowed to configure the time range, it is always set to 31 days
- [ ] A premium user is allowed to configure the time range
- [ ] If a premium user changes their time range, and then downgrades, the time range will be reset to 31 days at next login
- [ ] when loading a mail list while offline, only mail data that falls within the configured time range is available. All of those mails are totally available - mailbody, mailheaders, attachments list (but not the attachments or inline images themselves)
- [ ] When only part of a mail list successfully loaded, the retry button still shows.
- [ ] When coming back online, the retry button will successfully load the rest of the mail list
- [ ] trash and spam folders are always completely cleared
- [ ] only mail data is cleared: things like contacts and such will be available offline, regardless of the configured timerange
- [ ] when the offline data is cleared, the sqlite database itself should see a decrease in size (by default, sqlite doesn't shrink the database when doing `DELETE FROM`)
- [ ] #4079 

#### Regression tests
- [ ] loading of all lists. Lists should :
    - [ ] Contacts
    - [ ] Users
    - [ ] Templates
    - [ ] Search



##### Testing hints:
- [ ] You should test offline mode by actually disabling your network connection; some download requests (i.e. attachments) go via the native part, so turning off the network in dev tools will not correctly simulate being offline

**Pull request merged**